File: src/go/day2/main.go
package main

import (
	_ "embed"
	"fmt"
	"slices"

	"github.com/dancantos/advent2024/src/go/pkg/it"
	"github.com/dancantos/advent2024/src/go/pkg/timeit"
)

func main() {
	fmt.Printf("Puzzle1 (isSafe): %d\n", countSafe(Input, isSafe1))
	for range timeit.Run(100) {
		countSafe(Input, isSafe1)
	}

	fmt.Println()

	fmt.Printf("Puzzle2 (isSafeDampened): %d\n", countSafe(Input, isSafe2))
	for range timeit.Run(100) {
		countSafe(Input, isSafe2)
	}
}

const (
	decreasing = -1
	increasing = 1
)

func countSafe(input [][]int, safeFn func([]int) bool) int {
	return it.Count(it.Filter(slices.Values(input), safeFn))
}

File: src/python/day2/__main__.py
import modules.timer as timer
from .puzzle1 import safe1
from .puzzle2 import safe2
from .input import input

def run():
    print("puzzle1: count safe:", countSafe(input, safe1))
    for _ in timer.run(100):
        countSafe(input, safe1)

    print()

    print("puzzle2: count safe with dampening:", countSafe(input, safe2))
    for _ in timer.run(100):
        countSafe(input, safe2)

def countSafe(input, safe):
    counter = 0
    for v in input:
        if safe(v):
            counter+=1
    return counter

if __name__ == "__main__":
    run()